### DISCUSSION TITLE
Plugin doesn't starts rust-analyzer on Windows

### DISCUSSION THREAD

**Discussion by TheRustifyer** [mrcjkb/rustaceanvim#380] [Q&A]
### Have you read the docs and searched existing issues?

- [X] I have read [the README's troubleshooting guide](https://github.com/mrcjkb/rustaceanvim/blob/master/README.md#stethoscope-troubleshooting).
- [X] I have searched [existing issues](https://github.com/mrcjkb/rustaceanvim/issues?q=is%3Aopen+is%3Aissue+label%3Abug).
- [X] I have searched [existing discussions](https://github.com/mrcjkb/rustaceanvim/discussions).
- [X] I have searched the existing issues of other plugins related to this issue (if applicable).
- [X] I have searched the [existing rust-analyzer issues](https://github.com/rust-lang/rust-analyzer/issues?q=is%3Aissue+) (if applicable).
- [ ] I can reproduce the issue with stable Neovim (not a distribution or nightly) and the [minimal config](https://github.com/mrcjkb/rustaceanvim/blob/master/troubleshooting/minimal.lua).

### Neovim version (nvim -v)

v0.10.0-dev-3060+gd26943a6e

### Operating system/version

Windows 11

### Output of :checkhealth rustaceanvim

```console
.: require("..health").check()

- ERROR Failed to run healthcheck for "." plugin. Exception:
  [string "require("..health").check()"]:1: module '..health' not found:
  	no field package.preload['..health']
  cache_loader: module ..health not found
  cache_loader_lib: module ..health not found
  	no file '.\\\health.lua'
  	no file 'C:\Program Files (x86)\nvim\bin\lua\\\health.lua'
  	no file 'C:\Program Files (x86)\nvim\bin\lua\\\health\init.lua'
  	no file '.\\\health.dll'
  	no file 'C:\Program Files (x86)\nvim\bin\\\health.dll'
  	no file 'C:\Program Files (x86)\nvim\bin\loadall.dll'
  	no file '.\.dll'
  	no file 'C:\Program Files (x86)\nvim\bin\.dll'
  	no file 'C:\Program Files (x86)\nvim\bin\loadall.dll'
```

### How to reproduce the issue

```console
Install neovim on Windows 11
Install the required plugins
```

### Expected behaviour

It should be found, it is in the path and available. VSCode finds it.

### Actual behaviour

Plugin doesn't starts rust-analyzer

### The minimal config used to reproduce this issue.

```Lua
{
  'mrcjkb/rustaceanvim',
  version = '^4', -- Recommended
  ft = { 'rust' },
}
```

**Comment by mrcjkb** [comment]
Hey 👋 

Your `checkhealth` output tells be there's probably something wrong with your neovim installation.

What happens when you call it without the `rustaceanvim` arg?

**Reply by TheRustifyer** [reply]
Hey there @mrcjkb 

I have `nvim` built manually from today's latest commit.

I removed it and download the `v0.9.5` from `msys2 mingw64`.

This is the output of `checkhealth`:

```
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- OK [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap) installed.

Checking external dependencies ~
- OK rust-analyzer: found
- OK Cargo: found
- OK rustc: found
- OK debug adapter: found codelldb

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

In any case, `rust-analyzer` isn't being launched.

I don't install `ra` via *Mason*, since I typically rely on the default installation of `ra` via `rustup`, since I typically use both `Linux` and `Windows` environments along the day

Is this the expected way of configuring `rustaceanvim`?

Thanks for your attention

**Reply by mrcjkb** [reply]
🤔 the health check says it found the rust-analyzer wrapper, but it's not outputting a version, which is usually the case then the rustup component (`rustup component add rust-analyzer`) is missing.
Is your rustaceanvim installation up to date? In the latest version, the health check should give you an error if it can't determine the rust-analyzer version.

**Reply by TheRustifyer** [reply]
Sure. I reinstalled my plugins when I changed `nvim` to the `msys2` version, cleaned the `data` folder and let `lazy` reinstall everything again.
Nothing changed.

But maybe this:
```
rustup component add rust-analyzer
info: component 'rust-analyzer' for target 'x86_64-pc-windows-gnu' is up to date
```

Maybe the `gnu target` makes it fails for whatever reason?

**Reply by mrcjkb** [reply]
Can you run `:!rust-analyzer --version` in command mode?

**Reply by TheRustifyer** [reply]
Sorry for the late reply. I was on holidays.

rust-analyzer 1.78.0 (9b00956e 2024-04-29)

**Reply by mrcjkb** [reply]
Hmm, then it appears rust-analyzer is detected properly. Perhaps there's an error being thrown (e.g. by an ftplugin script or by another plugin) that is preventing the LSP client from starting. 

I recommend you follow the readme's troubleshooting guide, starting without any plugins or configs, and then gradually adding them until you can reproduce the behaviour.

**Reply by TheRustifyer** [reply]
Fine. Thanks for your help!

**Reply by TheRustifyer** [reply]
Hello again @mrcjkb 

I configured everything from scratch, and anything changed.

More notes:

- I have a dotfiles repo, that are shared with other Windows machine. The other machine starts `ra` without no problem
- I reinstalled `vscode` with the `ra` extension. Here's the logs:

```
2024-05-11T08:00:48.256186Z ERROR rust_analyzer::main_loop: FetchBuildDataError:
Failed to run build scripts for C:\Users\someone\code\zdc\Canyon-SQL: program not found
```

So, `vscode` is unable to work as well with `ra`. Something might be broken in my *Rust*  setup I guess.

In my Manjaro everything works fine.
In the other Windows machine, as I said, everything works fine too.

NOTE: `RustLsp` is not an editor command is shown when I manually try to trigger `:RustLsp <someaction>`

- I'll try to install the `msvc` based toolchain an try again
- I removed my `cargo.config` global cfg file, that was instructing Cargo to use `clang++ and the llvm-ar` as linker and archive.

Any other guess or happy idea?

Thanks again for your time

**Reply by mrcjkb** [reply]
I haven't touched a Windows machine in years, so I'm afraid I can't be of much help with that 😅

If it's working on one machine, but not on the other, it definitely looks like a toolchain or environment issue. 
I'd suggest looking at the [rust-analyzer troubleshooting guide](https://rust-analyzer.github.io/manual.html#troubleshooting), or asking in the [rust-analyzer discussions](https://github.com/rust-lang/rust-analyzer/discussions). You'll probably find more Windows users there.

**Reply by TheRustifyer** [reply]
It's perfectly fine. Thanks anyway for your time

**Reply by TheRustifyer** [reply]
Hey there. Sorry for bother you again.

I fix my `vscode` global json settings, because it wasn't working `ra` on `vscode` either.
After removing the override of `cargo check` for use `--all-features`, it came back to life on `vscode`.

Even tho, I went ahead and installed the `Mason` `ra` version, and received the warning pop-up.
I am using `lsp-zero` (I don't know if that can cause troubles) and removed the `rust-analyzer = lsp_zero.noop`,
and started to work again.

- I removed the `ra` installed from the *GNU* toolchain, left the `Mason` one. Neovim attaches.
- Removed the one from `Mason` and bring back the one of the *GNU* toolchain. Didn't attach.
- Bring back the one of `Mason`. Added again the `lsp_zero.noop`. Didn't attach.

So, `rustaceanvim` seems to not be able to attach the client to `ra`
I am only being able to get it work with `lsp-config` and the one from `rust-analyzer`.

I also installed the `msvc` *stable* toolchain, and didn't work at all.

Does this gave you some hint?

**Reply by TheRustifyer** [reply]
Same config on other Windows machine, and on a Manjaro. Can confirm that `rustaceanvim` works regardless of the underlying OS with my actual config.

Now on `Windows` have `msvc` as the default toolchain with the `ra` coming from `rustup`. No mason. Everything works "fine" with `lsp_zero`, but `rustaceanvim` doesn't start with any configuration

**Reply by mrcjkb** [reply]
> I am only being able to get it work with lsp-config and the one from rust-analyzer.

Interesting, so all those scenarios work with lspconfig, but not rustaceanvim?

If so, is this reproducible in both rustaceanvim and lspconfig with a minimal setup?

**Reply by TheRustifyer** [reply]
Sure. Exactly them same. It's kinda strange.

I cleaned my `nvim` installations from both Windows machines and tried again with the minimal configs. Anything changed.

What's more strange is that the plugin is loaded without problem, and commands even works too (for ex: `:RustEmitAsm`) but none of the related with `:RustLsp`. In concrete, the error is: `RustLsp, command not found`.

Is this telling you something different?

PD: Could you add on the checkhealth the `PATH` on where `rustaceanvim` found `ra`? Regarding my issue, it could be helpful for others that sometimes can face kind of similar issues.

**Reply by mrcjkb** [reply]
> Sure. Exactly them same. It's kinda strange.

What do you mean by this?

> What's more strange is that the plugin is loaded without problem, and commands even works too (for ex: :RustEmitAsm) but none of the related with :RustLsp. In concrete, the error is: RustLsp, command not found.

`RustEmitAsm` is a command created by Neovim's built-in rust filetype plugin, not rustaceanvim.
`RustLsp` commands are only created when the LSP client has attached. So it's to be expected that they don't work if your client isn't attaching.
The `RustAnalzyer start` command should be available regardless of lsp client status.

> Could you add on the checkhealth the PATH on where rustaceanvim found ra? 

The health check uses `vim.fn.executable`, which doesn't return the path.
On unix, you can run `:!which rust-analyzer` in command mode. I don't know what the equivalent command would be on Windows.

**Reply by TheRustifyer** [reply]
> What do you mean by this?

> Q: Interesting, so all those scenarios work with lspconfig, but not rustaceanvim?
If so, is this reproducible in both rustaceanvim and lspconfig with a minimal setup?

That with a minimal configuration only for `rustaceanvim` and other for `lspconfig` the issue is exactly the same

> On unix, you can run :!which rust-analyzer in command mode. I don't know what the equivalent command would be on Windows.

Since I am using a bash shell the execution of `ra` returns me the one bundled by default with the installation of `Rust` via `rustup`.

`Windows` equivalent is `where`. Same result.

I erased the `gcc` based toolchain (`pc-windows-gnu`) via `rustup`, in case it was causing conflicts (after installing the `pc-windows-msvc` one) in any way. Anything changed

**Reply by mrcjkb** [reply]
Maybe it's best if you set the command directly, via `vim.g.rustaceanvim.server.cmd` (see `:h rustaceanvim.config`).
